woocommerce/woocommerce-ios
# Persistence of sort order and filter options in the Product List

Fixes #2732

## Description
In the Product List, on Android, users are able to see the filters that they previously applied because on that platform we store the sort order and the filters information for products.
I implemented the same behavior on iOS. From now, if a user applies custom sort orders and filters in the Product List, now when they reopen the app will be able to see the previous settings applied.
For implementing this solution, after having received some [feedback from the team](https://github.com/woocommerce/woocommerce-ios/issues/2732#issuecomment-756820161), I implemented new methods in [AppSettingsStore](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/Yosemite/Stores/AppSettingsStore.swift) for fetching/upserting/deleting the product settings.
Why Product Settings (plural)? Since in the future probably we'll store other info about products, I decided to adopt a more general solution that can be reused for other purposes.
The Product Settings file is a `Plist` file stored locally, called `products-settings.plist`, that contains an array of `StoredProductSettings` entities, one for each `siteID`.

<img width="549" alt="Schermata 2021-01-11 alle 18 22 45" src="https://user-images.githubusercontent.com/495617/104303639-a0a18900-54ca-11eb-994f-fb4a07abd60a.png">



## Changes
- Created a new model that conforms to `Codable` called `StoredProductSettings` that presents all the information that we want to store into the Product Settings file.
- Since we encode the product status, the stock status, and the product type, I updated the models `ProductType`, `ProductStockStatus` and `ProductStatus` to conform to `Codable`.
- Updated the `ProductsViewController` that now is able to update and sync the local product settings.
- Implemented new methods `loadProductsSettings`, `upsertProductsSettings` and `resetProductsSettings` in `AppSettingsStore` + Unit Tests.
- Since we need to store the Products sort order, now the enum `ProductsStortOrder` is of type `String`.

## Testing
1. Open the product list, then apply sorting and filters to your liking.
2. Close the app.
3. Re-open the app, and make sure that the product list is still filtered/sorted with the filters that you previously applied.
4. Change store, then apply new sorting and filters.
5. Return to your initial store. The filters should be still set.
6. Logout from the app.
7. Do a new login to the same account -> The filters/sorting options should be reset for all the stores.


Update release notes:
